### PR TITLE
SNOW-719900 Remove random alias for subqeury in joins

### PR DESCRIPTION
**Pull request by sfc-gh-stan**
Please answer these questions before submitting your pull requests. Thanks!

1. What GitHub issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   Fixes #SNOW-719900

2. Fill out the following pre-review checklist:

   - [x] I am adding a new automated test(s) to verify correctness of my new code
   - [ ] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am adding new credentials
   - [ ] I am adding a new dependency

3. Please describe how your code solves the related issue.

  Previously, dataframe/table joins will use random aliases for the left and right subquery results at query generation, which means users cannot take advantage of query [result cache](https://community.snowflake.com/s/article/Caching-in-Snowflake-Data-Warehouse). This PR removes that and adds a unit test.

### CODE REVIEW

File: tests/unit/test_dataframe.py [hunk 1]
```diff
@@ -180,3 +180,22 @@ def test_create_or_replace_temp_view_bad_input():
         "The input of create_or_replace_temp_view() can only a str or list of strs."
         in str(exc_info)
     )
+
+
+def test_same_joins_should_generate_same_queries():
```
Comment (sfc-gh-jdu) [inline tests/unit/test_dataframe.py:185]: You can just run it as an integration test? Also, could you check whether q1.join(q2) works?

File: src/snowflake/snowpark/_internal/analyzer/analyzer_utils.py [hunk 2]
```diff
@@ -479,8 +477,8 @@ def set_operator_statement(left: str, right: str, operator: str) -> str:
 def left_semi_or_anti_join_statement(
     left: str, right: str, join_type: JoinType, condition: str
 ) -> str:
-    left_alias = random_name_for_temp_object(TempObjectType.TABLE)
-    right_alias = random_name_for_temp_object(TempObjectType.TABLE)
+    left_alias = "SNOWPARK_TEMP_LEFT_RESULT"
```
Comment (sfc-gh-sfan) [inline src/snowflake/snowpark/_internal/analyzer/analyzer_utils.py:480]: Do we need to worry about potential ambiguous alias if multiple joins are involved?

File: src/snowflake/snowpark/_internal/analyzer/analyzer_utils.py [hunk 3]
```diff
@@ -517,8 +515,8 @@ def left_semi_or_anti_join_statement(
 def snowflake_supported_join_statement(
     left: str, right: str, join_type: JoinType, condition: str
 ) -> str:
-    left_alias = random_name_for_temp_object(TempObjectType.TABLE)
-    right_alias = random_name_for_temp_object(TempObjectType.TABLE)
+    left_alias = "SNOWPARK_TEMP_LEFT_RESULT"
```
Comment (sfc-gh-yixie) [inline src/snowflake/snowpark/_internal/analyzer/analyzer_utils.py:518]: Some queries may have many repeated query aliases if we use a fixed one. For instance, in the following code df7's query will have two SNOWPARK_TEMP_LEFT_RESULT as subquery aliases. How about removing the alias if we use this way.
```
df5 = df1 .join(df2)
df6 = df3.join(df4)
df7 = df5.join(df6)
```

Comment (sfc-gh-stan) [reply src/snowflake/snowpark/_internal/analyzer/analyzer_utils.py:518]: That's a great point, why do we need these subquery aliases? I don't see them being used or saved separately from the query string 🤔

File: src/snowflake/snowpark/_internal/analyzer/analyzer.py [hunk 4]
```diff
@@ -161,23 +161,29 @@ def analyze(
 
         if isinstance(expr, Like):
             return like_expression(
-                self.analyze(expr.expr, parse_local_name), self.analyze(expr.pattern)
+                self.analyze(expr.expr, parse_local_name),
+                self.analyze(expr.pattern, parse_local_name),
```
Comment (sfc-gh-yixie) [inline src/snowflake/snowpark/_internal/analyzer/analyzer.py:165]: This (and other changes in this PR) is already in the main branch. I'm wondering how it's shown as a change in this PR.

File: src/snowflake/snowpark/_internal/analyzer/analyzer_utils.py (hunk 3 above)
Comment (sfc-gh-stan) [reply src/snowflake/snowpark/_internal/analyzer/analyzer_utils.py:518]: I don't have a strong opinion here but I think a longer name with `SNOWPARK` in it is likely to be duplicated.

**Review by sfc-gh-sfan — COMMENTED** [review]
Can we have a test of nested joins, e.g. df1.join(df2).join(df3)?

**Review by sfc-gh-sfan — APPROVED** [review]

**Pull request merged**